core/src/main/java/de/danoeh/antennapod/core/storage/DBWriter.java

```diff
@@ -472,7 +476,28 @@ private static void removeQueueItemSynchronous(final Context context,
         }
         adapter.close();
         if (performAutoDownload) {
-            DBTasks.autodownloadUndownloadedItems(context);
+            // delay the start of auto download for
+            // #2652, use case "The current playing episode should be at the front of the queue".
+            // when an episode playback completes and AntennaPod advances to the next episode,
+            // the following happens:
+            // 1. player plays the next episode in the background
+            // 2. remove the current one from the queue
+            // 3. trigger auto download (here)
+            //
+            // At this point,
+            // - the next episode's playback (step 1) typically has not yet started,
+            // - if we run autodownload right away, there is nothing playing yet
+            // - thus downloaded episodes will be queued to the very front
+            // - once playback has started, the in-progress episode will not be in the front,
+            //   (right behind the episodes being downloaded),
+            //   defeating the settings (of keeping in-progress at the front)
+            //
+            // The solution here is to delay the auto download, so that when it starts
+            // the next episode playing should be in-progress
+            // 30+ seconds should be enough, even if the next episode is streamed over the internet.
+            new Handler(Looper.getMainLooper()).postDelayed(() -> {
+                DBTasks.autodownloadUndownloadedItems(context);
+            }, 35000);
```

Comment:
This looks like a rather hacky approach. What if the user closes the app after removing a queue item?

Reply:
I'm not quite happy with it either, but the alternatives have their own drawbacks. Feedback welcome.

Recall the sequence of events leading to the scenario [*], when an episode playback is complete:
1. player plays the next episode in the background
2. the app automatically removes the current one from the queue
3. trigger auto download: at this point, the next episode to be played may or may not be in-progress yet (it's done in background)

The workaround here is to introduce a wait long enough so that the next episode to be played would be in-progress. I'd agree it is ugly. However, it has the advantage of the fix is in the context of why the fix is needed (autodownload logic).

For the case user closes the app after removing a queue item: autodownload won't kick in. 
The behavior is similar to what happens now: all autodownload triggers won't survive if users closes the app. [**]
The delay introduced does make it more likely to happen, however.

---
Alternatives considered:

**I.** Before starting to play the next episode in background (step 1 above), make the next episode in progress, e.g., setting its `FeedMedia.position` to `1` .

Concerns: it'd require adding some apparently out-of-place / unrelated logic in `PlaybackService`. I think it is in [`LocalPSMP.endPlayback()`](https://github.com/AntennaPod/AntennaPod/blob/d7f0f95adaa2adf15d3231994dab2eb469d8dd14/core/src/main/java/de/danoeh/antennapod/core/service/playback/LocalPSMP.java#L955)

In addition to adding complexity to `PlaybackService`, I am also concerned such subtle changes could get lost in future `PlaybackService` changes.

One advantage for this approach is that it becomes practical to write an androidTest to verify the behavior (no long wait).

---
**II.** Makes the actual playback logic record sooner (in DB) that it is in-progress.  The approach would not fully solve the problem, as the actual playback is done in the background. There is no guarantee that the playback would start before the dequeue happens.

---
[*] the scenario assumes
- continuous playback is enabled
- autodownload is enabled
- either
  - there is space in episode cache
  - autodownload > episode cleanup is set to when not in queue, or after finishing (which will free up space in episode cache)

[**] The lone exception is for the autodownload trigger upon feed refresh, as feed refresh is scheduled in a background job to begin with.

Reply:
Update: A way is found to resolve the issue without the above workaround.
- The latest commit uses a different (and more accurate) way to identify currently playing media [*]. 
- With it, it worked out that the by the time auto download is invoked as part of removeQueueItem, the next episode's state has been synchronously updated, so there is no more timing issue. [**]
- A end-to-end androidTest will be added, as the behavior is subjected to the internals of `PlaybackService`. The to-be-added test will help to ensure future changes in `PlaybackService` won't break it.

---
[*] The latest commit uses the state in `PlaybackPreferences`, rather than relying on `PlayMedia.position` in db.

[**] The new code path that worked out is as follows
1. When episode playback is complete
2. In `thread-A`, `LocalPSMP.onEndPlayback()` is invoked
3. it at some point plays the next media, writing the new status in `PlaybackPreferences` synchronously in the same thread.
4. it then calls `onPostPlayback()`, which indirectly triggers auto download via removeQueueItem